rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py

```diff
@@ -1,203 +1,135 @@
-        pi_outs[Columns.ACTIONS] = (prior_action, posterior_action)
-
-        return pi_outs
-
     @override(TorchRLModule)
     def _forward_inference(self, batch: Dict[str, TensorType]) -> Dict[str, TensorType]:
-        # Encoder forward pass.
-        encoder_out = self.encoder(batch)
-        # Policy head forward pass.
-        return self.pi(encoder_out[ENCODER_OUT], inference=True)
+        return self._pi(batch[Columns.OBS], inference=True)
 
     @override(TorchRLModule)
     def _forward_exploration(
         self, batch: Dict[str, TensorType], **kwargs
     ) -> Dict[str, TensorType]:
-        # Encoder forward pass.
-        encoder_out = self.encoder(batch)
-        # Policy head forward pass.
-        return self.pi(encoder_out[ENCODER_OUT], inference=False)
+        return self._pi(batch[Columns.OBS], inference=False)
 
     @override(TorchRLModule)
     def _forward_train(self, batch: Dict[str, TensorType]) -> Dict[str, TensorType]:
         return self._forward_exploration(batch)
 
     @override(ValueFunctionAPI)
     def compute_values(self, batch: Dict[str, TensorType], embeddings=None):
-        # Encoder forward pass to get `embeddings`, if necessary.
-        if embeddings is None:
-            embeddings = self.encoder(batch)[ENCODER_OUT]
-        # Value head forward pass.
-        vf_out = self.vf(embeddings)
+        # Value function forward pass.
+        vf_out = self._value_net(batch[Columns.OBS])
         # Squeeze out last dimension (single node value head).
         return vf_out.squeeze(-1)
+
+    def _pi(self, obs, inference: bool):
+        # Prior forward pass.
+        prior_out = self._prior_net(obs)
+        dist_a1 = TorchCategorical.from_logits(prior_out)
+
+        # If in inference mode, we need to set the distribution to be deterministic.
+        if inference:
+            dist_a1 = dist_a1.to_deterministic()
+        # Sample a1.
+        a1 = dist_a1.sample()
+
+        # Posterior forward pass.
+        posterior_batch = torch.cat(
+            [obs, one_hot(a1, self.action_space[0])],
+            dim=-1,
+        )
+        posterior_out = self._posterior_net(posterior_batch)
+        dist_a2 = TorchDiagGaussian.from_logits(posterior_out)
+        if inference:
+            dist_a2 = dist_a2.to_deterministic()
+
+        a2 = dist_a2.sample()
+
+        actions = (a1, a2)
+
+        # We need the log-probabilities for the loss.
+        outputs = {
+            Columns.ACTION_LOGP: (
+                TorchMultiDistribution((dist_a1, dist_a2)).logp(actions)
+            ),
+            Columns.ACTION_DIST_INPUTS: torch.cat([prior_out, posterior_out], dim=-1),
+            # Concatenate the prior and posterior actions and log probabilities.
+            Columns.ACTIONS: actions,
+        }
+
+        return outputs
+
+    @override(TorchRLModule)
+    def get_inference_action_dist_cls(self):
+        return TorchMultiDistribution.get_partial_dist_cls(
```

Comment:
This is a nice example of how to use our `MultiDistribution` classes. I wonder, however, why we do not use it above where the distributions are used for sampling? In training I guess exactly this distribution is used for the action logps, isn't it?

Reply:
It's still a very messy API:
* `from_logits` requires a single tensor, which then gets split into n sub-tensors through the help of another method 🤔 
* `ctor` however utilizes a nested structure of subcomponents, which should probably be the norm. I'm wondering, whether we can simplify the entire `Distributions` API further by getting rid of `from_logits` altogether.
* Another, even more radical thought would be to even get rid of `Distribution/TorchDistribution` and use `torch.Distribution` instead 😉.

But we should clean this up separately.